src/basic/unit-file.c

```diff
@@ -261,6 +261,86 @@ static int directory_name_is_valid(const char *name) {
         return false;
 }
 
+static int unit_file_resolve_symlink(
+                const char *root_dir,
+                char **search_path,
+                const char *dir,
+                int dirfd,
+                const char *filename,
+                char **ret_simplified,
+                const char **ret_dst) {
+
+        _cleanup_free_ char *target = NULL, *simplified = NULL;
+        const char *dst;
+        int r;
+
+        assert(dir);
+        assert(dirfd >= 0);
+        assert(filename);
+
+        r = readlinkat_malloc(dirfd, filename, &target);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to read symlink %s%s%s: %m",
+                                         dir, dir ? "/" : "", filename);
+
+        const bool is_abs = path_is_absolute(target);
+        if (root_dir || !is_abs) {
+                char *target_abs = path_join(is_abs ? root_dir : dir, target);
+                if (!target_abs)
+                        return log_oom();
+
+                free_and_replace(target, target_abs);
+        }
+
+        /* Get rid of "." and ".." components in target path */
+        r = chase_symlinks(target, root_dir, CHASE_NOFOLLOW | CHASE_NONEXISTENT, &simplified, NULL);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to resolve symlink %s/%s pointing to %s: %m",
+                                         dir, filename, target);
+
+        /* Check if the symlink goes outside of our search path.
+         * If yes, it's a linked unit file or mask, and we don't care about the target name.
+         * Let's just store the link source directly.
+         * If not, let's verify that it's a good symlink. */
+        char *tail = path_startswith_strv(simplified, search_path);
+        if (!tail) {
+                log_debug("%s: linked unit file: %s/%s → %s",
+                          __func__, dir, filename, simplified);
+                dst = filename;
+
+        } else {
+                bool self_alias;
+
+                dst = basename(simplified);
```

Comment:
if you touch this, maybe also use path_extract_filename()...

Reply:
We know that the name is "well formed", and basename works nicely. It also avoid the need to pass a separate allocated pointer around and free it.

Reply:
I'd really prefer to add basename() to the list of functions that LGTM and coccinelle will complain/rewrite. The semantics are just too weird I think, and it's always good to have additional checks in place that the assumptions about "well-formedness" actually hold...